src/graph/transform/knn.cc

```diff
@@ -0,0 +1,117 @@
+
+    // create view for each segment
+    NDArray current_data_points = const_cast<NDArray*>(&data_points)->CreateView(
+      {d_length, feature_size}, data_points->dtype, d_offset * feature_size * sizeof(FloatType));
+    FloatType * current_query_pts_data = query_points_data + q_offset * feature_size;
+
+    KDTreeNDArrayAdaptor<FloatType, IdType> kdtree(feature_size, current_data_points);
+
+    // query
+#pragma omp parallel for
+    for (int64_t q = 0; q < q_length; ++q) {
+      std::vector<IdType> out_buffer(k);
+      std::vector<FloatType> out_dist_buffer(k);
+      auto curr_out_offset = 2 * k * q + out_offset;
+      FloatType * q_point = current_query_pts_data + q * feature_size;
+      size_t num_matches = kdtree.GetIndex()->knnSearch(
+        q_point, k, out_buffer.data(), out_dist_buffer.data());
+
+      for (size_t i = 0; i < num_matches; ++i) {
+        result_data[curr_out_offset] = q + q_offset; curr_out_offset++;
+        result_data[curr_out_offset] = out_buffer[i] + d_offset; curr_out_offset++;
+      }
+    }
+  }
+}
+}  // namespace impl
+
+template <DLDeviceType XPU, typename FloatType, typename IdType>
+void KNN(const NDArray data_points, const IdArray data_offsets,
+         const NDArray query_points, const IdArray query_offsets,
+         const int k, IdArray result) {
+  impl::KdTreeKNN<FloatType, IdType>(
+    data_points, data_offsets, query_points, query_offsets, k, result);
+}
+
+template void KNN<kDLCPU, float, int32_t>(
+  const NDArray data_points, const IdArray data_offsets,
+  const NDArray query_points, const IdArray query_offsets,
+  const int k, IdArray result);
+template void KNN<kDLCPU, float, int64_t>(
+  const NDArray data_points, const IdArray data_offsets,
+  const NDArray query_points, const IdArray query_offsets,
+  const int k, IdArray result);
+template void KNN<kDLCPU, double, int32_t>(
+  const NDArray data_points, const IdArray data_offsets,
+  const NDArray query_points, const IdArray query_offsets,
+  const int k, IdArray result);
+template void KNN<kDLCPU, double, int64_t>(
+  const NDArray data_points, const IdArray data_offsets,
+  const NDArray query_points, const IdArray query_offsets,
+  const int k, IdArray result);
+
+DGL_REGISTER_GLOBAL("transform._CAPI_DGLKNN")
+.set_body([] (DGLArgs args, DGLRetValue *rv) {
+    const NDArray data_points = args[0];
+    const IdArray data_offsets = args[1];
+    const NDArray query_points = args[2];
+    const IdArray query_offsets = args[3];
+    const int k = args[4];
+    IdArray result = args[5];
+
+    aten::CheckContiguous(
+      {data_points, data_offsets, query_points, query_offsets, result},
+      {"data_points", "data_offsets", "query_points", "query_offsets", "result"});
+    aten::CheckCtx(
+      data_points->ctx, {data_offsets, query_points, query_offsets, result},
+      {"data_offsets", "query_points", "query_offsets", "result"});
+
+    ATEN_XPU_SWITCH_CUDA(data_points->ctx.device_type, XPU, "KNN", {
```

Comment:
Seems that we don't support CUDA for now.  Change to `ATEN_XPU_SWITCH` so it will fail on GPU tensors.